### ISSUE TITLE
Boolean lists can be optimized as bitset/bitvecs

### ISSUE THREAD

**Issue by abacef** [rust-lang/rust-clippy#8893]
### What it does

Reading the README of the [bitvec](https://crates.io/crates/bitvec) crate, it says

> To get started, you can perform basic text replacement on your project. Translate any existing types as follows:
[bool; N] becomes BitArray
[bool] becomes BitSlice
Vec<bool> becomes BitVec
Box<[bool]> becomes BitBox

I was wondering how easy it would be to write a performance lint for this since I know at least in C, each bool is a usize long. I am not sure if the Rust compiler optimizes this code to make each bool a bit and packs them together. There are many bitvec-like crates that people could use.

### Lint Name

bool_collection

### Category

perf

### Advantage

Bits take less space than bools, and are probably faster to work with.

### Drawbacks

I am sure there are tons of false positives where it is useful to keep an array of booleans. Also, bitvec is a crate, so another dependency will need to be added to the crate/binary. Also bitvec is not the newest but not the matureist either, and I think this is the most mature bit manipulation crate.

### Example

Using the [bit_array](https://crates.io/crates/bit-array) crate
```rust
fn main() {
  let mut bool_list = [false; 4];
  bool_list[1] = true;
  
  let sum = bool_list.iter().filter(|x| **x).count();
}
```

Could be written as:

```rust
extern crate typenum;
use bit_array::BitArray;
use typenum::{Unsigned, U4};

fn main() {
  let mut bitarr = BitArray::<u32, U4>::from_elem(false);
  bitarr.set(1, true);
  
  let sum = bitarr.iter().filter(|x| *x).count();
}
```

**Comment by y21** [comment]
It's also not possible to get a `&bool` from a bitvec, so this is something to keep in mind if someone wants to implement this.
So, the following shouldn't get linted:
```rs
fn takes_ref(_: &bool) {}

let v = [false];
takes_ref(&v[0]);
```

**Comment by Centri3** [comment]
It's also worth mentioning

> and are probably faster to work with.

Is not entirely true. The smallest bit of information the CPU knows is a byte, to get a single bit requires extra instructions to isolate it from the byte (at least, `bt`). On its own, it's a good chunk slower.

That's it, though. pretty neat idea!
